In [4]:
import pandas as pd
from transformers import DistilBertTokenizer
import kagglehub
path = kagglehub.dataset_download("ashfakyeafi/spam-email-classification")
print("Path to dataset files:", path)

/Users/marksmacbookair/miniconda/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset files: /Users/marksmacbookair/.cache/kagglehub/datasets/ashfakyeafi/spam-email-classification/versions/3


In [6]:
import opendatasets as od 
import pandas 

od.download( 
	"https://www.kaggle.com/datasets/ashfakyeafi/spam-email-classification/data") 

Skipping, found downloaded files in "./spam-email-classification" (use force=True to force download)


In [7]:
import kagglehub
import pandas as pd

# Download the dataset
dataset_path = kagglehub.dataset_download("ashfakyeafi/spam-email-classification")

file_path = f"{dataset_path}/email.csv"

emails = pd.read_csv(file_path)

# Print the first 5 records
print("First 5 records:\n", emails.head())

First 5 records:
   Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...


In [8]:
import json
emails.columns = emails.columns.str.strip().str.lower()

# Open a file to write JSONL format
with open('spam_ham_dataset.jsonl', 'w', encoding='utf-8') as outfile:
    for _, row in emails.iterrows():
        entry = {
            "messages": [
                {"role": "user", "content": row["message"]},
                {"role": "assistant", "content": row["category"]}
            ]
        }
        json.dump(entry, outfile)
        outfile.write('\n')



In [ ]:
import openai
openai.api_key = '********************'


In [11]:
file = openai.files.create(
  file=open("spam_ham_dataset.jsonl", "rb"),
  purpose='fine-tune'
)
print(file.id)


file-BM9QsgTit66wGmf44HrFfx


In [13]:
fine_tune_response = openai.fine_tuning.jobs.create(
    training_file='file-HfPLP3dYpBxC7yuZX2Zgfb',
    model='gpt-3.5-turbo'
)

print(fine_tune_response)


FineTuningJob(id='ftjob-QP9nklaffp7P9Pvy6Vwgdf7s', created_at=1743375011, error=Error(code=None, message=None, param=None), fine_tuned_model=None, finished_at=None, hyperparameters=Hyperparameters(batch_size='auto', learning_rate_multiplier='auto', n_epochs='auto'), model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-kl35lFMDQ0CAqAfBpssxQbOj', result_files=[], seed=878374318, status='validating_files', trained_tokens=None, training_file='file-HfPLP3dYpBxC7yuZX2Zgfb', validation_file=None, estimated_finish=None, integrations=[], metadata=None, method=Method(dpo=None, supervised=MethodSupervised(hyperparameters=MethodSupervisedHyperparameters(batch_size='auto', learning_rate_multiplier='auto', n_epochs='auto')), type='supervised'), user_provided_suffix=None)


In [43]:
#Checking the progress in fine tuning.
openai.fine_tuning.jobs.retrieve("ftjob-B6SnKRT9PeUA1gt764jyqp1O")


FineTuningJob(id='ftjob-B6SnKRT9PeUA1gt764jyqp1O', created_at=1743195117, error=Error(code=None, message=None, param=None), fine_tuned_model='ft:gpt-3.5-turbo-0125:personal::BGBqjRQe', finished_at=1743198495, hyperparameters=Hyperparameters(batch_size=11, learning_rate_multiplier=2.0, n_epochs=3), model='gpt-3.5-turbo-0125', object='fine_tuning.job', organization_id='org-kl35lFMDQ0CAqAfBpssxQbOj', result_files=['file-H85EB9SEoNDghzUsm4Qfs2'], seed=205853128, status='succeeded', trained_tokens=541497, training_file='file-HfPLP3dYpBxC7yuZX2Zgfb', validation_file=None, estimated_finish=None, integrations=[], metadata=None, method=Method(dpo=None, supervised=MethodSupervised(hyperparameters=MethodSupervisedHyperparameters(batch_size=11, learning_rate_multiplier=2.0, n_epochs=3)), type='supervised'), user_provided_suffix=None)

In [68]:
# Testing!
response = openai.chat.completions.create(
    model='ft:gpt-3.5-turbo-0125:personal::BGBqjRQe',
    messages=[
        {"role": "user", "content": "You won a $1000 gift card, click here!"}
    ],
    max_tokens=1,
    temperature=0
)

classification = response.choices[0].message.content.strip()
print(f"Classification: {classification}")  # Should output: "spam" or "ham"


Classification: spam


### Accuracy Overview
Now that I've confirmed that the function works and outputs something, it is time to put a confusion matrix and measure the accuracy of the model after finetuning.

In [69]:
from sklearn.metrics import confusion_matrix, classification_report
import re

true_labels = []
predicted_labels = []
# prompt = f"""Classify the following SMS message as ham or spam, and return the probability of it being spam.
# Message: '{message}'
# Output only: ham or spam, and the probability with two decimal points.
# """
for i in range(100): 
    
    random_number = random.randint(1, len(emails))
    true_label = emails["category"][random_number]
    message = emails["message"][random_number]
    
    response = openai.chat.completions.create(
    model='ft:gpt-3.5-turbo-0125:personal::BGBqjRQe',
    messages=[
        {"role": "user", "content": message}
    ],
    max_tokens=1,
    temperature=0
)

    # Extract predicted label
    predicted_label = str(response.choices[0].message.content.strip())

    true_labels.append(true_label)
    predicted_labels.append(predicted_label)

# Generate confusion matrix
cm = confusion_matrix(true_labels, predicted_labels, labels=["ham", "spam"])
print("Confusion Matrix:\n", cm)

# classification report
print("\nClassification Report:")
print(classification_report(true_labels, predicted_labels, labels=["ham", "spam"]))


Confusion Matrix:
 [[88  0]
 [ 1 11]]

Classification Report:
              precision    recall  f1-score   support

         ham       0.99      1.00      0.99        88
        spam       1.00      0.92      0.96        12

    accuracy                           0.99       100
   macro avg       0.99      0.96      0.98       100
weighted avg       0.99      0.99      0.99       100



# Confusion Matrix Breakdown:

Ham: Out of 88 ham messages, all were correctly classified (88 true negatives, 0 false positives).

Spam: Out of 12 spam messages, 11 were correctly identified as spam and only 1 was misclassified as ham (1 false negative).

Performance Metrics:

Ham: Precision is 0.99, and recall is 1.00, which means the model is almost perfect in identifying ham messages.

Spam: Precision is perfect (1.00), indicating that every message predicted as spam was indeed spam. However, the recall is 0.92, meaning 8% of spam messages were missed.

Overall Accuracy: The model achieves 99% accuracy, with both macro and weighted averages above 0.96, reflecting very high performance across both classes.

Conclusion:
The model performs nearly flawlessly with an overall accuracy of 99%. The only slight shortfall is the one misclassified spam message, which lowers the spam recall to 92%. This could be a focus for further refinement, but overall, the model is highly effective for the task.

## Shocking result of Current LLM Model
I tried just using the current ChatGPT model (GPT 3.5) without training and fine tuning to classify spam/ham email content and the results were shocking. 

In [63]:
message = emails.iloc[random.randint(1, len(emails))]
from openai import OpenAI

client = OpenAI(
    api_key = "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",
)
prompt = f"""Classify the following sms message as ham or spam. Also return the probability of it being spam.
Message: '{message}'.
The output should only contain two words: ham or spam, and the probability with two decimal points.
"""

print(prompt)

response = client.chat.completions.create(
  model="gpt-3.5-turbo",
  temperature = 0.0,
  messages=[
        {"role": "user", "content": prompt}
    ]
)

print("=============================")
print("This email is " + response.choices[0].message.content)


Classify the following sms message as ham or spam. Also return the probability of it being spam.
Message: 'category                                           ham
message     Im good! I have been thinking about you...
Name: 4822, dtype: object'.
The output should only contain two words: ham or spam, and the probability with two decimal points.

This email is ham, 0.01


In [61]:

true_labels = []
predicted_labels = []

for i in range(100): 
    random_number = random.randint(1, len(emails))
    true_label = emails["category"][random_number]
    message = emails["message"][random_number]

    prompt = f"""Classify the following sms message as ham or spam. 
Message: '{message}'
Output only: ham or spam
              """

    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        temperature=0.0,
        messages=[{"role": "user", "content": prompt}]
    )

    # Extract predicted label
    predicted_label = str(response.choices[0].message.content)

    true_labels.append(true_label)
    predicted_labels.append(predicted_label)

# Generate confusion matrix
cm = confusion_matrix(true_labels, predicted_labels, labels=["ham", "spam"])
print("Confusion Matrix:\n", cm)

# classification report
print("\nClassification Report:")
print(classification_report(true_labels, predicted_labels, labels=["ham", "spam"]))


Confusion Matrix:
 [[81  2]
 [ 1 16]]

Classification Report:
              precision    recall  f1-score   support

         ham       0.99      0.98      0.98        83
        spam       0.89      0.94      0.91        17

    accuracy                           0.97       100
   macro avg       0.94      0.96      0.95       100
weighted avg       0.97      0.97      0.97       100



# Confusion Matrix Breakdown:

Ham (Actual 83):

81 correctly classified as ham (true negatives).

2 misclassified as spam (false positives for spam).

Spam (Actual 17):

16 correctly classified as spam (true positives).

1 misclassified as ham (false negative).

Performance Metrics:

Ham:

Precision: 0.99, meaning nearly all messages predicted as ham are indeed ham.

Recall: 0.98, so 98% of actual ham messages were correctly identified.

F1-Score: 0.98, reflecting excellent balance between precision and recall.

Spam:

Precision: 0.89, which indicates that about 11% of messages predicted as spam are actually ham.

Recall: 0.94, meaning the model caught 94% of actual spam messages.

F1-Score: 0.91, showing strong performance, though slightly lower than for ham.

## Overall Accuracy:
The model achieves a 97% accuracy on the dataset, with both macro and weighted averages indicating robust performance.

## Conclusion:
The classifier performs very well overall, with only minor misclassifications. While ham messages are almost perfectly recognized, there’s a small drop in spam precision, suggesting a few ham messages are mistakenly flagged as spam. This performance is strong, but further tuning might help to balance the precision-recall trade-off for spam if that’s critical for your application.


# Comparison of the Two Confusion Matrices
Both reports yield overall accuracies above 97%, indicating that the model is reliably distinguishing between ham and spam.

### Class-Specific Performance:

Ham messages: The model rarely misclassifies ham, with near-perfect performance.

Spam messages: While one misclassification is common in both reports, the precision and recall remain strong, although the second report shows a minor drop in spam precision (0.89), suggesting a few ham messages might be incorrectly flagged as spam.

Surprising Effectiveness of a Non-Finetuned Model:

Unexpected Competence:
Despite not being fine-tuned for spam detection, the regular ChatGPT-3.5 model demonstrates very high accuracy. This is surprising because one might expect a model specialized on this task to significantly outperform a general-purpose language model.

Broad Training Data:
The model’s impressive performance can be attributed to its extensive pre-training on a diverse dataset, which includes a wide range of text genres and patterns. This broad exposure allows it to implicitly learn features useful for distinguishing spam from ham.

Robust Language Understanding:
ChatGPT's strong natural language understanding capabilities enable it to pick up on subtle cues in the messages that differentiate spam from legitimate messages, even without explicit fine-tuning.

Final Thoughts
The high performance of the regular ChatGPT-3.5 model on this classification task is both encouraging and a testament to the power of large-scale language models. However, keep in mind:

Consistency:
Performance might vary with different prompt formulations or datasets.

Fine-Tuning Potential:
While the base model already performs remarkably well, targeted fine-tuning could further optimize accuracy, especially for edge cases or more nuanced spam messages.

In summary, these results highlight how versatile and robust modern LLMs are, even for specialized tasks like spam detection, despite not being explicitly trained for that purpose.